File: Updater/converters/DerivedBidiClass.rb
=begin
 
 DerivedBidiClass.rb
   ©︎ 2018 YOCKOW.
     Licensed under MIT License.
     See "LICENSE.txt" for more information.
 
=end


module DerivedBidiClass
  PATH = 'Sources/UnicodeSupplement/__data__/UCD/DerivedBidiClass.swift'
  URLs = [
    URI.parse('https://www.unicode.org/Public/UCD/latest/ucd/extracted/DerivedBidiClass.txt')
  ]
end

module DerivedBidiClass; class << self
  def write(remote_io, file)
    normal_lines = []
    comment_lines = []
    
    remote_io.each {|line|
      next if line =~ /^\s*$/
      if line =~ /^\s*\u{23}\s*(.+)$/
        comment_lines.push($1.strip)
      else
        normal_lines.push(line)
      end
    }
    
    self.write_normal(normal_lines, file)
    self.write_default(comment_lines, file)
  end
  
  def write_normal(lines, file)
    table = Table.new(lines)
    failed("No data about BidiClass") if table.count < 1
    
    file.puts(UnicodeTable.new(table).to_swift_assoc_array('bidiClass', 'String'))
  end
  
  def write_default(lines, file)
    default_value_ranges = {}
    default_value_properties = {}
    type_ranges = ''
    type_properties = ''
    
    lines.each {|line|
      if line =~ /^The\s+unassigned\s+code\s+points\s+that\s+default\s+to\s+([A-Z]+)\s+(.+)/
        bc = $1
        st = $2
        if st =~ /range/
          type_ranges = bc
        elsif st =~ /propert(y|ies)/
          type_properties = bc
        else
          failed("Unexpected line: #{line}")
        end
      elsif type_ranges != ''
        default_value_ranges[type_ranges] = [] if line =~ /^\[/
        default_value_ranges[type_ranges].concat(line.gsub(/^\[/, '').gsub(/\]$/, '').split(/\s+/))
        type_ranges = '' if line =~ /\]$/
      elsif type_properties != ''
        default_value_properties[type_properties] = [] if !default_value_properties[type_properties]
        if line =~ /^[A-Z_a-z]+$/
          default_value_properties[type_properties].push(line)
        else
          type_properties = ''
        end
      end
    }
    
    ## RANGES
    default_value_ranges_table_source = []
    default_value_ranges.keys.each {|bc|
      default_value_ranges[bc].each{|range_string|
        converted_range_string = range_string.gsub(/\\[Uu]/, '').gsub(/\-/, '..')
        default_value_ranges_table_source.push("#{converted_range_string}; #{bc}")
      }
    }
    
    default_value_ranges_table = Table.new(default_value_ranges_table_source)
    file.puts(UnicodeTable.new(default_value_ranges_table).
              to_swift_assoc_array('bidiClass_defaultValue_ranges', 'String'))
    
    
    ## PROPERTIES
    array_type = '[(KeyPath<Unicode.Scalar.LatestProperties,Bool>, Bool, String)]'
    file.write("internal let _bidiClass_defaultValue_properties:#{array_type} = [")
    default_value_properties.keys.each{|bc|
      default_value_properties[bc].each{|property|
        computed_property_name = 'is' + property.to_upper_camel_case
        file.write("(\\.#{computed_property_name},true,\"#{bc}\"),")
      }
    }
    file.puts("]")
  end
end; end

File: utils/SwiftUnicodeSupplementUpdater/Sources/UnicodeSupplementUpdater/Delegates/DerivedBidiClass.swift
/* *************************************************************************************************
 DerivedBidiClass.swift
   © 2020 YOCKOW.
     Licensed under MIT License.
     See "LICENSE.txt" for more information.
 ************************************************************************************************ */
 
import Foundation
import Ranges
import StringComposition
import UnicodeSupplement
import yCodeUpdater
import yExtensions

extension StringProtocol {
  /// e.g. "\\u0600" -> "\u{0600}"
  private var _scalarValue: Unicode.Scalar.Value {
    assert(self.lowercased().hasPrefix("\\u"))
    return Unicode.Scalar.Value(self.dropFirst(2), radix: 0x10)!
  }
  
  /// e.g. "\\u0600-\\u07BF" -> "\u{0600}"..."\u{07BF}"
  fileprivate var _scalarValueRange: ClosedRange<Unicode.Scalar.Value> {
    let components = self.split(separator: "-")
    assert(components.count == 2)
    return components[0]._scalarValue...components[1]._scalarValue
  }
  
  fileprivate var _scalarValueRanges: [ClosedRange<Unicode.Scalar.Value>] {
    let rangeStrings = self.split(omittingEmptySubsequences: true) { $0.isNewline || $0.isWhitespace }
    return rangeStrings.compactMap {
      var string = $0
      if string.hasPrefix("[") { string = string.dropFirst() }
      if string.hasSuffix("]") { string = string.dropLast() }
      return string._scalarValueRange
    }
  }
}

open class DerivedBidiClass: UCDPropertiesCodeUpdaterDelegate<Unicode.BidiClass> {
  open override var prefix: String {
    return "bidiClass"
  }
  
  open override var sourceURLs: Array<URL> {
    return [
      URL(string: "https://www.unicode.org/Public/UCD/latest/ucd/extracted/DerivedBidiClass.txt")!
    ]
  }
  
  open override func reduce(columns: [String]) throws -> Unicode.BidiClass {
    return Unicode.BidiClass(abbreviated: columns.first!)!
  }
  
  open override func describe(value: Unicode.BidiClass) -> String {
    return ".\(String(describing: value))"
  }
  
  private enum _DefaultValue {
    case ranges([ClosedRange<Unicode.Scalar.Value>])
    case properties([String])
    
    fileprivate mutating func add(_ string: String) {
      switch self {
      case .ranges(var array):
        array.append(contentsOf: string._scalarValueRanges)
        self = .ranges(array)
      case .properties(var array):
        array.append(string)
        self = .properties(array)
      }
    }
  }
  open override func convert<S>(_ intermediates: S) throws -> StringLines where S: Sequence, S.Element == IntermediateDataContainer<UnicodeData> {
    let intermediates = Array(intermediates)
    
    assert(intermediates.count == 1)
    // Extract special default values in comments
    var defaultValue: _DefaultValue? = nil
    var key: Substring? = nil
    var defaultValues: [String: _DefaultValue] = [:]
    let prefix = "The unassigned code points that default to "
    let rangesSuffix = " are in the ranges:"
    let rangeSuffix = " are in the range:"
    let propertiesSuffix = " have one of the following properties:"
    for row in intermediates.first!.content.rows {
      // Break out the loop when comment lines end.
      if row.payload != nil { break }
      guard let comment = row.comment else { continue }
      if comment.hasPrefix(prefix) {
        assert(defaultValue == nil && key == nil)
        if comment.hasSuffix(rangesSuffix) || comment.hasSuffix(rangeSuffix) {
          defaultValue = .ranges([])
        } else if comment.hasSuffix(propertiesSuffix) {
          defaultValue = .properties([])
        }
        assert(defaultValue != nil, "Unexpected comment line: \(comment)")
        key = comment.dropFirst(prefix.count).splitOnce(separator: " ").0
      } else if key != nil {
        assert(defaultValue != nil)
        if comment.isEmpty {
          defaultValues[String(key!)] = defaultValue!
          key = nil
          defaultValue = nil
        } else {
          defaultValue!.add(comment)
        }
      }
    }
    assert(key == nil && defaultValue == nil)
    
    var defaultValueRanges = RangeDictionary<Unicode.Scalar.Value, Unicode.BidiClass>()
    var defaultValueProperties = Dictionary<String, [String]>()
    for (key, defaultValue) in defaultValues {
      switch defaultValue {
      case .ranges(let ranges):
        for range in ranges {
          defaultValueRanges.insert(Unicode.BidiClass(abbreviated: key)!, forRange: AnyRange(range))
        }
      case .properties(let properties):
        defaultValueProperties[key] = properties
      }
    }
    
    var rangeDictionary = defaultValueRanges
    for (range, value) in try self._convert(intermediates) {
      rangeDictionary.insert(value, forRange: range)
    }
    
    var result: StringLines = self._convert(rangeDictionary, describer: self.describe(value:))
    
    do { // Default Properties
      result.append("// Default Values defined by core properties")
      
      let keyPathOriginalTypeName = "(KeyPath<Unicode.Scalar.LatestProperties, Bool>, Bool, Unicode.BidiClass)"
      let keyPathTypeName = self.typeAliasName(for: keyPathOriginalTypeName)
      func _tripleID(_ nx: Int) -> String {
        return "__\(self.prefix)_default_properties_triple_\(nx._base36)"
      }
      func _triple(_ prop: String, _ status: Bool, _ bidiClass: Unicode.BidiClass) -> String {
        return "(\\.is\(prop.upperCamelCase), \(status ? "true" : "false"), \(self.describe(value: bidiClass)))"
      }
      
      var nn = 0
      for (bidiString, properties) in defaultValueProperties.sorted(by: { $0.key < $1.key }) {
        let bidiClass = Unicode.BidiClass(abbreviated: bidiString)!
        for property in properties {
          result.append("private let \(_tripleID(nn)): \(keyPathTypeName) = \(_triple(property, true, bidiClass))")
          nn += 1
        }
      }
      result.append("internal let _\(self.prefix)_default_properties: Array<\(keyPathOriginalTypeName)> = [")
      for ii in 0..<nn {
        result.append(String.Line("\(_tripleID(ii)),", indentLevel: 1)!)
      }
      result.append("]")
    }
    
    return result
  }
}
